### ISSUE TITLE
Error loading extensions with shortcuts that have default value

### ISSUE THREAD

**Issue by benvarkey** [ipython-contrib/IPython-notebook-extensions#388]
I am on a Mac, using Safari. When using extensions that have keyboard shortcuts with default values (e.g. `toggle_all_line_numbers`), I get an error loading the extension. 
I chased it down to 
 **nbextensions/config/hotkey_editor.js**: **humanize_shortcut**: 
`_.map(shortcut.split('-'), this.humanize_key ).join(joinchar);`

FYI, replacing that line with `shortcut.split('-').join(joinchar);` fixes the problem, but I don't know what `_.map` was used for.

**Comment by jcb91** [comment]
OK, thanks for the report. I think the `_.map` call is simply to apply
`humanize_shortcut` to each key in the combination, which should do some
replacements like '-' for 'minus', and on mac, things like 'alt' for the
appropriate unicode key symbol if I recall correctly. Presumably I've made
a boo-boo somewhere there, so I'll take a look today.

**Comment by jcb91** [comment]
OK, so I've discovered/thought about a few things.

First, some of the extensions use the actions framework to register their keyboard shortcuts. The fact that the notebook recently changed from using a period to using a colon as the separator between the action prefix and action name mean that some of the extensions end up throwing errors while attempting to add button groups, if you're on a new enough version of the notebook (anything after   7df94eb0013f9990b85c380e453728b8c18a38ca). This is enough to stop the relevant extensions from loading correctly. The extensions in question are autoscroll, comment-uncomment, init_cell, read-only, and toggle_all_line_numbers. 

Second, I'm now a little confused by what exactly you're reporting, so I have a few question:

* What version of jupyter are you using?

* From the issue title

  > Error loading extensions with shortcuts that have default value

  I assume that you mean that if you set a specific value, the extension works correctly, but if you try to keep the default, it doesn't. Is that correct?

* When you say 

  > I get an error loading the extension

  I assume you mean you've seen an error message of some kind in the browser javascript console. Is that right? Could you post (all of) the console output from loading a regular notebook?

* You say you've tracked it down to the `humanize_shortcut ` function in `nbextensions/config/hotkey_editor.js`. What lead you there? The `nbextensions/config/hotkey_editor.js` file should (I think) only be used when configuring the shortcut from the `/nbextensions` URL - it isn't actually used in the other extensions themselves such as `toggle_all_line_numbers`, which are loaded when you open a notebook. I have a feeling that what you're describing in that file might be a separate bug related to whether the humanize_key method is provided by the quickhelp module. Again, the version of notebook that you're using might help to identify this.

* On a Mac, any shortcut of the form alt-letter is liable to not work correctly when any text input element (such as a notebook cell) has focus, since the keypress is interpreted as the beginning of typing a special character (e.g. alt-n,n gives ñ). These alt-based shortcuts seem to still work in command mode, but honestly this kind of behaviour is why I wrote the hotkey_editor in the first place!

So, yes, let me know if any of that helps/doesn't make sense/whatever.

Best

J

**Comment by benvarkey** [comment]
# Updated error report

I should have been clearer about the problem, let me try again.

Extensions with hotkeys--*Codefolding*, *Comment/Uncomment Hotkey* and *Toggle all line numbers*--that have a `default` key under the `Parameters` in the *.yaml* file, render in the *.../nbextensions* page with an error `error loading extension Toggle all line numbers`:
<img width="1095" alt="screen shot 2015-11-01 at 5 45 30 pm" src="https://cloud.githubusercontent.com/assets/2020374/10872893/6daab8d4-80c0-11e5-86c1-384c93b59ca0.png">

The console shows the following:
<img width="750" alt="screen shot 2015-11-01 at 5 57 25 pm" src="https://cloud.githubusercontent.com/assets/2020374/10872986/31627040-80c2-11e5-9036-71e7d7638b9b.png">
<img width="751" alt="screen shot 2015-11-01 at 5 57 49 pm" src="https://cloud.githubusercontent.com/assets/2020374/10872989/35bf0e96-80c2-11e5-8456-3db17661d392.png">
<img width="751" alt="screen shot 2015-11-01 at 5 58 14 pm" src="https://cloud.githubusercontent.com/assets/2020374/10872990/3a199e5c-80c2-11e5-9822-8447a79d7b04.png">

If I comment out just the `default` key in the *.yaml* file, the page renders without any issues:
<img width="1039" alt="screen shot 2015-11-01 at 5 51 31 pm" src="https://cloud.githubusercontent.com/assets/2020374/10872931/4048bf0c-80c1-11e5-8ab0-13ffcade3fbf.png">

However, irrespective of the error, the extension **is** loaded (if activated) with the toolbar button, 
when a notebook is launched.

Once I had the console error messages, I traced the error back to:
- *main.js* (616): `build_params_ui(extension.Parameters)`
- *main.js* (687): `set_input_value(input, param['default']);`
- *main.js* (149): `.html(hke.humanize_shortcut(new_value))`
- *hotkey_editor.js* (54): `var sh = _.map(shortcut.split('-'), this.humanize_key ).join(joinchar);`

# Reply
Now, to answer your questions:

* Jupyter version: 4.0.6
* I can't change the default value because the page doesn't render properly.
* I have posted the relevant sections from the javascript console.
* Yeah, looks like `humanize_key` method is missing.
* :-)

**Comment by jcb91** [comment]
Ah, ok, that clears a few things up, I'd somehow got stuck on it being an issue with individual extensions, but in retrospect, it should have been obvious to me that it's the config system. And yes, it seems like I've just missed out one last patch function which should have been provided - the humanize_key method...

**Comment by jcb91** [comment]
should hopefully be fixed by #389

**Comment by jcb91** [comment]
whoops, spoke too soon, will need another tweak :S

**Comment by jcb91** [comment]
OK, I think that works for me now, could you let me now whether #389 fixes things for you, when you get a chance? Thanks!
J

**Comment by benvarkey** [comment]
That works. Thanks!

**Comment by jfbercher** [comment]
Super work @jcb91!
